airbyte-integrations/connectors/source-linnworks/source_linnworks/streams.py

```diff
@@ -0,0 +1,227 @@
+
+
+class StockLocations(LinnworksStream):
+    # https://apps.linnworks.net/Api/Method/Inventory-GetStockLocations
+    # Response: List<StockLocation> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockLocation
+    # Allows 150 calls per minute
+    primary_key = "StockLocationIntId"
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Inventory/GetStockLocations"
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        records = super().read_records(sync_mode, cursor_field, stream_slice, stream_state)
+
+        for record in records:
+            location = Location(authenticator=self.authenticator)
+            srecords = location.read_records(sync_mode, cursor_field, stream_slice, {"pkStockLocationId": record["StockLocationId"]})
+            record["location"] = next(srecords)
+            yield record
+
+
+class StockItems(LinnworksStream):
+    # https://apps.linnworks.net//Api/Method/Stock-GetStockItemsFull
+    # Response: List<StockItemFull> https://apps.linnworks.net/Api/Class/linnworks-spa-commondata-Inventory-ClassBase-StockItemFull
+    # Allows 250 calls per minute
+    primary_key = "StockItemIntId"
+    page_size = 200
+
+    raise_on_http_errors = False
+
+    def path(
+        self, stream_state: Mapping[str, Any] = None, stream_slice: Mapping[str, Any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> str:
+        return "/api/Stock/GetStockItemsFull"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        url = urlparse(response.request.url)
+        qs = dict(parse_qsl(url.query))
+
+        page_size = int(qs.get("entriesPerPage", self.page_size))
+        page_number = int(qs.get("pageNumber", 0))
+
+        data = response.json()
+
+        if response.status_code == requests.codes.ok and len(data) == page_size:
+            return {
+                "entriesPerPage": page_size,
+                "pageNumber": page_number + 1,
+            }
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], stream_slice: Mapping[str, any] = None, next_page_token: Mapping[str, Any] = None
+    ) -> MutableMapping[str, Any]:
+        params = {
+            "entriesPerPage": self.page_size,
+            "pageNumber": 1,
+            "loadCompositeParents": "true",
+            "loadVariationParents": "true",
+            "dataRequirements": "[0,1,2,3,4,5,6,7,8]",
+            "searchTypes": "[0,1,2]",
+        }
+
+        if next_page_token:
+            params.update(next_page_token)
+
+        return params
+
+
+class IncrementalLinnworksStream(LinnworksStream, ABC):
```

Comment:
A pattern of having `Incremental{source_name}Stream` class is followed in virtually all of the sources. Event with one concrete class, e.g. `IncrementalLeverHiringStream`, `IncrementalDixaStream`. Do you want me to discard the pattern? Of course, it's a trivial change, and it would take less time than writing this comment. But I think that consistency is of more value here.

For now, `stream_slices` and `get_updated_state` operate solely on ` start_date` and `cursor_field`. Both variables are quite generic and might be reused for additional streams.

Reply:
Update:

I've moved `stream_slices` in response to [this comment](https://github.com/airbytehq/airbyte/pull/7588#discussion_r745879382) to `ProcessedOrders` since it's become stream-specific. However `get_updated_state` still looks pretty generic.